pkg/io/grpc.go

```diff
@@ -253,20 +255,36 @@ loop:
 }
 
 func (s *StreamIOReader) Read(p []byte) (n int, err error) {
-	for b := range s.bytes {
-		nb := copy(p, b)
-		return nb, nil
+	var b []byte
+	ok := true
+
+	if len(s.buffer) > 0 { // if bytes in buffer, use it first
+		b = s.buffer
+	} else {
+		b, ok = <-s.bytes
+	}
+	nb := copy(p, b)
+
+	if nb < len(b) {
+		s.buffer = b[nb:] // if bytes left, store in buffer
+	} else {
+		s.buffer = nil
 	}
-	return 0, io.EOF
+
+	if !ok {
+		return nb, io.EOF
+	}
+
+	return nb, nil
 }
 
 func (s *StreamIOReader) WriteTo(w io.Writer) (n int64, err error) {
 	for b := range s.bytes {
 		nb, err := w.Write(b)
+		n += int64(nb)
 		if err != nil {
 			return n, err
 		}
```

Comment:
so it can error and still read some bytes successfully?

Reply:
Yes, some implementations might do that. It’s mentioned in the interface.